### PR TITLE
Add a pre-connection check to the home and siteurl

### PR DESCRIPTION
**Pull request by georgestephanis**
This should rule out ones that map to private or reserved ranges, and
provide better error messages.

We regularly get folks trying to connect and hitting errors against these rules on wpcom -- so may as well try the rules on the .org site.

My only concern is if gethostbyname() somehow points to a local ip address because of some /etc/hosts rule on the server, it may somehow shut it down.  I'm not sure how likely that is, though, so we may want to rewrite it a touch for a softer launch, or include some sort of override.

### CODE REVIEW

**Comment by georgestephanis** [comment]
It may benefit from some unit tests on specific domain names that would map to private or reserved ip ranges.

File: class.jetpack-data.php [hunk 1]
```diff
@@ -39,4 +39,87 @@ public static function get_access_token( $user_id = false ) {
...
+	 * @param $domain
+	 * @param array $extra
+	 *
+	 * @return bool|WP_Error
+	 */
+	public static function is_usable_domain( $domain, $extra = array() ) {
+
+		// If it's empty, just fail out.
+		if ( ! $domain ) {
+			return new WP_Error( 'fail_domain_empty', sprintf( 'Domain `%1$s` just failed is_usable_domain check as it is empty.', $domain ) );
```
Comment (jeherve) [inline class.jetpack-data.php]: Shouldn't we make all these strings translateable by adding the `jetpack` textdomain?

**Comment by dereksmart** [comment]
Good idea.  The only other thing I can think of is that gethostbyname() could <a href="http://php.net/manual/en/function.gethostbyname.php#104944">take a while</a> resolving the DNS for some hosts, especially if run twice. 
Though my tests showed it was pretty quick (average ~0.02 seconds).

File: class.jetpack-data.php (hunk 1 above)
Comment (georgestephanis) [inline class.jetpack-data.php]: I couldn't recall offhand what the default was for error messages, but thinking back, yeah probably.

**Comment by georgestephanis** [comment]
It should only run twice if the first time fails, or the siteurl and home have two different domains.

**Comment by georgestephanis** [comment]
Also, this will only run just as they're trying to connect, so it won't be a common occurence.

**Comment by samhotchkiss** [comment]
Is this still `in progress`, as the tag would indicate?

File: class.jetpack.php [hunk 2]
```diff
@@ -2298,6 +2298,21 @@ public static function unlink_user( $user_id = null ) {
 	 * Attempts Jetpack registration.  If it fail, a state flag is set: @see ::admin_page_load()
 	 */
 	public static function try_registration() {
+		// Let's get some testing in beta versions and such.
+		if ( self::is_development_version() ) {
+			// Before attempting to connect, let's make sure that the domains are viable.
+			$domains_to_check = array_unique( array(
+				'siteurl' => parse_url( get_site_url(), 'host' ),
```
Comment (dereksmart) [inline class.jetpack.php]: `'host'` is causing this to throw errors.  I think we want `PHP_URL_HOST` here

Comment (georgestephanis) [inline class.jetpack.php]: Drat, yeah.

**Comment by dereksmart** [comment]
LGTM :+1: :100: :zap:

**Pull request merged**